File: question/neetcode150/trees/balanced-binary-tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        ans = True

        def dfs(node: Optional[TreeNode]):
            nonlocal ans
            if not node:
                return 0
            
            left = dfs(node.left)
            right = dfs(node.right)
            if abs(left-right) > 1:
                ans = False

            return 1 + max(left,right)
        
        dfs(root)
        return ans

File: question/neetcode/binary_tree/balanced-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, bool& valid){
        if(not node) return 0;

        int left = dfs(node->left, valid);
        int right = dfs(node->right, valid);

        if(abs(left-right)>1) valid = false;

        return 1+max(left,right);
    }
    bool isBalanced(TreeNode* root) {
        bool valid = true;
        dfs(root, valid);

        return valid;
    }
};